### DISCUSSION TITLE
Inconsistency in bedrock data between chtexture>`texcl` & corestrictions>`resdept`

### DISCUSSION THREAD

**Discussion by kevinwolz** [ncss-tech/soilDB#323] [Q&A] [answered]
To identify depth to bedrock (in finer detail than was is offered via muaggatt>`brockdepmin`), I have been evaluating two differnet approaches:

1. Evaluating chtexture>`texcl` across all horizons, search for the topmost horizon with a value of "Bedrock", and take the chorzon>`hzdept_r`
2. Finding a corestrictions>`resdept` at which corestrictions>`reskind` has a value of "Bedrock". (I originally ignored this approach as I had been thinking that "restrictive layers" were everything _other than_ bedrock, but I guess not!)

Unfortunately, it seems like these two approaches are not consistent with each other. Specifically, it seems like (A) anytime the texture approach identifies Bedrock, so does Approach 2, but (B) there are many mukeys for which Approach 2 identifies Bedrock but Approach 1 does not.
![image](https://github.com/ncss-tech/soilDB/assets/11711566/f8dcceea-bfb1-4117-9a1e-2128f16774c1)

What do I trust!?

That last row also has a weird inconsistency between the `*_l` variables, but I'm not sure how widespread this is.

**Comment by kevinwolz** [comment]
Okay I just (partially) answered by own question. It turns out that, while I was accounting for variations on the "bedrock" descriptor in the corestrictions approach, I was not including "Unweathered bedrock" or "Weathered bedrock" in the texture approach.

After including those, here is the new comparison. Still not completely identical, but certainly better. Because it appears that there is better utilization of the h/r/l spread in the corestrictions approach, my gut says that approach is better. But there's that one mukey that  has bedrock in the texture approach but not the corestrictions approach!

![image](https://github.com/ncss-tech/soilDB/assets/11711566/fd6c8670-5d4f-437d-9485-3cf7ad978943)

**Reply by brownag** [reply, marked as answer]
Good, I was going to just point out that there are two other obsolete "lieutex" (in lieu texture) codes used for bedrock in the database. "Bedrock" corresponds to `BR` but you also have "Unweathered bedrock" `UWB` and "Weathered bedrock" `WB` still in common use across the database.

As far as inconsistencies in depths there is has not been a strict requirement that the low-rv-high for component horizon depth v.s. component restriction depth match exactly.  Most component horizons in SSURGO do not have their depth range populated anyway, only the RV. I assume that (chorizon) is where your `texture_bedrock_*` depth columns come from; depth range by layer is still not required for our minimum data standard but is generally recommended these days. 

It is rare that we have sufficient data to really customize the restriction depths beyond nominal depth class ranges e.g. "moderately deep 50-100cm" but there are commonly few-centimeter differences due to unit conversion from inches->centimeters, the addition of an O horizon, or similar. 

L-RV-H is a requirement for corestrictions, though, so if you need that range then it is wise to use that route. 

142cm in your edge case: I would note that there may be a different, perhaps shallower, non-bedrock restriction? Again consider the choice list options for `reskind` as you needed to for `lieutex`. There is always the possibility of slight inconsistencies, especially in older components that have not been through a more recent update with our more modern battery of validations.

``` r
library(soilDB)
x <- SDA_query("SELECT TOP 1000 component.cokey, reskind, resdept_l, resdept_r, resdept_h, 
             MIN(hzdept_r) AS hzbedrockdept_r FROM component
           LEFT JOIN corestrictions ON component.cokey = corestrictions.cokey AND reskind LIKE '%bedrock%'
           INNER JOIN chorizon ON component.cokey = chorizon.cokey 
           INNER JOIN chtexturegrp ON chorizon.chkey = chtexturegrp.chkey 
           INNER JOIN chtexture ON chtexturegrp.chtgkey = chtexture.chtgkey
           WHERE lieutex IN ('Weathered bedrock', 'Unweathered bedrock', 'Bedrock')
           GROUP BY component.cokey, reskind, resdept_l, resdept_r, resdept_h")
#> single result set, returning a data.frame

mismatches <- subset(x, resdept_r != hzbedrockdept_r)
nrow(mismatches)
#> [1] 177

x$resdept_c <- cut(x$resdept_r, c(50, 100, 150, 1e5))
x$hzbedrockdept_c <- cut(x$hzbedrockdept_r, c(50, 100, 150, 1e5))

# ~10x fewer differ in terms of depth class
class_mismatches <- subset(x, resdept_c != hzbedrockdept_c)
nrow(class_mismatches)
#> [1] 20
```

**Reply by brownag** [reply]
RE: mukey 3257822
``` r
library(soilDB)
x <- SDA_query("SELECT mapunit.mukey, component.cokey, component.compname, reskind, resdept_l, resdept_r, resdept_h, 
             MIN(hzdept_l) AS hzbedrockdept_l, 
             MIN(hzdept_r) AS hzbedrockdept_r,
             MIN(hzdept_h) AS hzbedrockdept_h FROM mapunit
           INNER JOIN component ON mapunit.mukey = component.mukey AND mapunit.mukey = 3257822
           LEFT JOIN corestrictions ON component.cokey = corestrictions.cokey AND reskind LIKE '%'
           INNER JOIN chorizon ON component.cokey = chorizon.cokey 
           INNER JOIN chtexturegrp ON chorizon.chkey = chtexturegrp.chkey 
           INNER JOIN chtexture ON chtexturegrp.chtgkey = chtexture.chtgkey
           WHERE lieutex IN ('Weathered bedrock', 'Unweathered bedrock', 'Bedrock')
           GROUP BY mapunit.mukey, component.cokey, component.compname, reskind, resdept_l, resdept_r, resdept_h")
#> single result set, returning a data.frame
x
#>     mukey    cokey     compname            reskind resdept_l resdept_r
#> 1 3257822 24434525      Gaphill Paralithic bedrock       102       142
#> 2 3257822 24434526      Gaphill Paralithic bedrock        NA       142
#> 3 3257822 24434529    Rockbluff Paralithic bedrock        NA       132
#> 4 3257822 24434524 Rock outcrop Paralithic bedrock         0         0
#>   resdept_h hzbedrockdept_l hzbedrockdept_r hzbedrockdept_h
#> 1       203              NA             142              NA
#> 2        NA              NA             142              NA
#> 3        NA              NA             132              NA
#> 4         0              NA               0              NA
```

**Reply by kevinwolz** [reply]
Thank you for the additional detail on this! "Paralithic bedrock" is definitely my vocab word for the day...
